packages/tpetra/core/src/Tpetra_CrsSingletonFilter_LinearProblem_def.hpp

```diff
@@ -0,0 +1,919 @@
+  auto RowMapColors_Data            = RowMapColors_->getLocalViewHost(Tpetra::Access::ReadWrite);
+  auto ColMapColors_Data            = ColMapColors_->getLocalViewHost(Tpetra::Access::ReadWrite);
+
+  for (int i = 0; i < localNumRows; i++) {
+    // Get ith row
+    GetRow(i, NumIndices, localIndices);
+    for (size_t j = 0; j < NumIndices; j++) {
+      local_ordinal_type ColumnIndex = localIndices[j];
+
+      // Bounds check for ColumnIndex
+      if (static_cast<size_t>(ColumnIndex) >= ColProfilesData.extent(0)) {
+        std::cout << "Error: ColumnIndex out of bounds: " << ColumnIndex << std::endl;
+        std::abort();
+      }
+
+      ColProfilesData(ColumnIndex, 0)++;  // Increment column count
+
+      if (static_cast<size_t>(ColumnIndex) >= localRowIDofSingletonColData.extent(0)) {
+        std::cout << "Error: ColumnIndex out of bounds for localRowIDofSingletonColData: "
+                  << ColumnIndex << std::endl;
+        std::abort();
+      }
+
+      // Record local row ID for current column
+      // will use to identify row to eliminate if column is a singleton
+      localRowIDofSingletonColData(ColumnIndex, 0) = i;
+    }
+    // If row has single entry, color it and associated column with color=1
+    if (NumIndices == 1) {
+      int j2 = localIndices[0];
+      ColHasRowWithSingletonData(j2, 0)++;
+      RowMapColors_Data(i, 0)  = 1;
+      ColMapColors_Data(j2, 0) = 1;
+      localNumSingletonRows_++;
+    }
+  }
+
+  // 1) The vector ColProfiles has column nonzero counts for each processor's contribution
+  // Combine these to get total column profile information and then redistribute to processors
+  // so each can determine if it is the owner of the row associated with the singleton column
+  // 2) The vector ColHasRowWithSingleton[i] contain count of singleton rows  that are associated with
+  // the ith column on this processor.  Must tell other processors that they should also eliminate
+  // these columns.
+
+  // Make a copy of ColProfiles for later use when detecting columns that disappear locally
+  vector_type_int NewColProfiles(ColProfiles.getMap());
+  NewColProfiles.update(1.0, ColProfiles, 0.0);
+  auto NewColProfilesData = NewColProfiles.getLocalViewHost(Tpetra::Access::ReadWrite);
+
+  // If importer is non-trivial, we need to perform a gather/scatter to accumulate results
+  auto importer = FullCrsMatrix()->getCrsGraph()->getImporter();
+  if (importer != Teuchos::null) {
+    vector_type_int tmpVec(FullMatrixDomainMap());  // Use for gather/scatter of column vectors
+    tmpVec.putScalar(0);
+    tmpVec.doExport(ColProfiles, *importer, ADD);
+    ColProfiles.doImport(tmpVec, *importer, INSERT);
+
+    tmpVec.putScalar(0);
+    tmpVec.doExport(ColHasRowWithSingleton, *importer, ADD);
+    ColHasRowWithSingleton.doImport(tmpVec, *importer, INSERT);
+  }
+
+  for (size_t j = 0; j < NumIndices; j++) {
```

Comment:
Same `parallel_reduce` pattern.

Reply:
I think @iyamazaki has addressed this by implementing run_on_host_ and parallel_reduce options for testing.